### ISSUE TITLE
M2.2.8 to M2.3.3 Upgrade Wishlist Duplicate Key Name

### ISSUE THREAD

**Issue by nickpiro** [magento/magento2#26501]
Attempting to upgrade our Magento 2.2.8 to 2.3.3 yields various errors. When running setup:upgrade, this SQL error pops up:
SQLSTATE[42000]: Syntax error or access violation: 1061 Duplicate key name 'WISHLIST_CUSTOMER_ID', query was: ALTER TABLE `wishlist` ADD CONSTRAINT `WISHLIST_CUSTOMER_ID` UNIQUE KEY (`customer_id`)

### Preconditions (*)
Magento 2.3.3 upgraded from 2.2.8

### Steps to reproduce (*)
1. Backup composer.json file: cp composer.json composer.json.bak
2. composer require magento/product-community-edition=2.3.3 --no-update
3. composer require --dev allure-framework/allure-phpunit:~1.2.0 friendsofphp/php-cs-fixer:~2.14.0 lusitanian/oauth:~0.8.10 magento/magento-coding-standard:~3.0.0 magento/magento2-functional-testing-framework:2.4.5 pdepend/pdepend:2.5.2 phpmd/phpmd:@stable phpunit/phpunit:~6.5.0 sebastian/phpcpd:~3.0.0 squizlabs/php_codesniffer:~3.4.0 --sort-packages --no-update
4. composer remove --dev sjparkinson/static-review fabpot/php-cs-fixer --no-update
5. composer update
6. php bin/magento cache:clean
7. Manually clean var/cache, var/page_cache, generated/code/*
8. php bin/magento setup:upgrade

### Expected result (*)

1. Magento install, minimal issues.

### Actual result (*)

1. When running setup:upgrade, this SQL error pops up:
SQLSTATE[42000]: Syntax error or access violation: 1061 Duplicate key name 'WISHLIST_CUSTOMER_ID', query was: ALTER TABLE `wishlist` ADD CONSTRAINT `WISHLIST_CUSTOMER_ID` UNIQUE KEY (`customer_id`)

**Comment by engcom-Bravo** [comment]
Hello @nickpiro

Thank you for your report.

We had similar issues reported before. It is not always reproducible, that is why we can't start fixing it.
I'm not able to reproduce this issue now.

I can offer you a workaround that should help. Rename the existing index for the wishlist table in your database:
```alter table wishlist rename index WISHLIST_CUSTOMER_ID to WISHLIST_CUSTOMER_ID_2; ```

and perform ```php bin/magento setup:upgrade```

Please, let us know if it helped.

Thank you in advance.

**Comment by nickpiro** [comment]
@engcom-Bravo would I rename the wishlist table before or after the upgrade?

**Comment by nickpiro** [comment]
Also, we're using MySQL 5.6. The syntax above works for 5.7 and above!

**Comment by engcom-Bravo** [comment]
@nickpiro 

Please, try to use this query 
```
ALTER TABLE `wishlist` CHANGE COLUMN `customer_id` `customer_id2` int unsigned default 0 not null comment 'Customer ID';
```
And then, try to run ```php bin/magento setup:upgrade```

Thank you

**Comment by nickpiro** [comment]
Hey @engcom-Bravo that worked in regards to MySQL, but this is the message I received when Schema creation/updates:
Syntax error or access violation: 1061 Duplicate key name 'WISHLIST_CUSTOMER_ID', query was: ALTER TABLE `wishlist` ADD COLUMN `customer_id` int(10) UNSIGNED NOT NULL DEFAULT 0  COMMENT "Customer ID", ADD CONSTRAINT `WISHLIST_CUSTOMER_ID_CUSTOMER_ENTITY_ENTITY_ID` FOREIGN KEY (`customer_id`) REFERENCES `customer_entity` (`entity_id`)  ON DELETE CASCADE, ADD CONSTRAINT `WISHLIST_CUSTOMER_ID` UNIQUE KEY (`customer_id`)

**Comment by nickpiro** [comment]
Do you think there is something in the wishlist dbschema that causes it not to ignore the fact that the wishlist_customer_id already exists?

**Comment by engcom-Bravo** [comment]
@nickpiro 

Great that it helped.

Turned out that there has been a fix related to this behavior 
![DeepinScreenshot_select-area_20200203141618](https://user-images.githubusercontent.com/51680745/73652699-25b3d980-4690-11ea-879b-fda88adc02e8.png)

This issue will not appear on Magento 2.3.4

We have to close this issue.

Thank you for collaboration

**Comment by nickpiro** [comment]
I too found those changes and made them but they did not correct the issue. I'm still receiving the duplicate key name issue.

**Comment by nickpiro** [comment]
@engcom-Bravo this issue should not be closed. I am still not able to upgrade to 2.3+ due to this issue!

**Comment by nickpiro** [comment]
It seems the setup/Schema/AddProductIdConstraint.php is not included in 2.3.4, but is in 2.3.3. There are also changes that were done to db_schema_whitelist.json but none of them seem to make a difference with my issue that I am seeing.

Should the setup/Schema/AddProductIdConstraint.php be used? Or the updates to the db_schema_whitemist.json? 

I need help with this!

**Comment by nickpiro** [comment]
I think the issue lies somewhere in here:

<constraint xsi:type="foreign" referenceId="WISHLIST_CUSTOMER_ID_CUSTOMER_ENTITY_ENTITY_ID" table="wishlist"
                    column="customer_id" referenceTable="customer_entity" referenceColumn="entity_id"
                    onDelete="CASCADE"/>
        <constraint xsi:type="unique" referenceId="WISHLIST_CUSTOMER_ID">
            <column name="customer_id"/>
        </constraint>

**Comment by nickpiro** [comment]
Just for anyone else who has this issue: I've been able to resolve it by removing the following from the db_schema.xml
`<constraint xsi:type="unique" referenceId="WISHLIST_CUSTOMER_ID">
            <column name="customer_id"/>
        </constraint>`

Also, the setup/Schema/AddProductIdConstraint.php is not used in 2.3.0.

**Comment by operator888** [comment]
Hi @nickpiro ,
I just stumbled upon the same error after working on a magento2.3.5 branch installing the Pearl theme and then I switched back to master branch and can not run the magento setup:upgrade because of the same error.
And in my db_schema.xml there is no record about the wishlist. 
Do you have any suggestion...?
Thank you

**Comment by seattletechworks** [comment]
I recently upgraded my store from Magento 2.3.5p1 to 2.4.1. I had also removed WeltPixel's Pearl theme. At least, I thought it was completely removed. During the upgrade, while running "php bin/magento setup:upgrade", I encountered the error -

SQLSTATE[42000]: Syntax error or access violation: 1061 Duplicate key name 'WISHLIST_CUSTOMER_ID', query was: ALTER TABLE wishlist ADD CONSTRAINT WISHLIST_CUSTOMER_ID UNIQUE KEY (customer_id)

I traced it back to this module - WeltPixel_AdvancedWishlist. Try disabling it first - "php bin/magento module:disable WeltPixel_AdvancedWishlist" and then rerun your upgrade.
